llvm/lib/CodeGen/ScheduleDAGInstrs.cpp

```diff
@@ -263,6 +275,9 @@ void ScheduleDAGInstrs::addPhysRegDataDeps(SUnit *SU, unsigned OperIdx) {
       bool ImplicitPseudoUse = false;
       SDep Dep;
       if (UseOpIdx < 0) {
+        // FIXME: UseOpIdx can be passed to computeOperandLatency, which can
+        //   pass it to findUseIdx, which treats it as unsigned. If this is
+        //   the expected behavior, it should be commented.
```

Comment:
Comment is weird, should just fix this to be consistent

Reply:
I'm not sure how to fix this.
For the reference, here is how the passed `UseOpIdx` is used in `findUseIdx()`:
```
static unsigned findUseIdx(const MachineInstr *MI, unsigned UseOperIdx) {
  unsigned UseIdx = 0;
  for (unsigned i = 0; i != UseOperIdx; ++i) {
    const MachineOperand &MO = MI->getOperand(i);
    if (MO.isReg() && MO.readsReg() && !MO.isDef())
      ++UseIdx;
  }
  return UseIdx;
}
```
This code should crash if `UseOperIdx == -1`, except it doesn't. This function is only called (under certain conditions) if the target uses SchedModel (not Itineraries). Probably this code is unreachable for in-tree models.

Naively fixing this by setting `ImplicitPseudoUse` to true results in too many test failures (> 650) and crashes Hexagon quite badly.

Reply:
Should I reword the comment, perhaps? I think this still deserves a FIXME.

Reply:
Operand indexes are unsigned. Where is the signed usage?

Adding a comment that asks a question about fixing a comment doesn't help the reader of the comment

Reply:
> Operand indexes are unsigned. Where is the signed usage?

Here: https://github.com/llvm/llvm-project/blob/8b87e2ffcf597d7eba1cf3ff12bea1fae1f999e6/llvm/lib/CodeGen/ScheduleDAGInstrs.cpp#L277

Reply:
I tried to improve this comment, please see if it makes more sense now.

Reply:
> This code should crash if `UseOperIdx == -1`, except it doesn't. This function is only called (under certain conditions) if the target uses SchedModel (not Itineraries). Probably this code is unreachable for in-tree models.

I did some debugging. It appears that when -1 is passed to `computeOperandLatency`, `UseInstr` is always nullptr. Operand index is only used if `UseInstr` is non-null, so strictly speaking there is no bug. I'm going to drop this comment.
(Let me know if I should do something different, like add a note or assertion somewhere.)